### DISCUSSION TITLE
Can Triber take advantage of a forwarded port in NAT?

### DISCUSSION THREAD

**Discussion by vi** [Tribler/tribler#6912] [General]
Typically P2P apps have a configurable port and overridable external address, so that they can accept incoming connections more easily.

In Tribler's Settings, I see only 2 ports in "Connection" tab: a port of a proxy (unapplicable) and a port to use between core and GUI. Does it mean that Tribler don't need a forwarded port, even if one exists in otherwise NAT-ed connection?

**Comment by qstokkink** [comment]
It is _usually_ not necessary to set up port forwarding for Tribler. However, if you want to make use of a forwarded port, you can set a value for the `port` under the `libtorrent` section in your configuration file.

https://github.com/Tribler/tribler/blob/2a96c4e0422b0a3bf9c1d1d3245296e0cfb42adb/src/tribler/core/components/libtorrent/settings.py#L21

The external address is not configurable and is always set to `0.0.0.0`. You would have to change Tribler's source code here to set that:

https://github.com/Tribler/tribler/blob/1609e8d85dc42c87db8fbe640404877ee1d2edf6/src/tribler/core/components/libtorrent/download_manager/download_manager.py#L249

**Reply by vi** [reply]
I don't see mentions of a configuration file in GUI "settings" window. Shall it contain a hint that there are more Tribler settings that are visibile in Settings window. Shall there be a system like Firefox's `about:config` to edit advanced settings?

**Reply by vi** [reply]
> The external address is not configurable and is always set to 0.0.0.0

By "external address" I mean the IP address Tribler requests other peers (especially the ones behind NAT) to connect to.

For example, it auto-detected NAT reflexive address is `1.2.3.4` internal IP Tribler is running on is `192.168.0.2` and there is a port forward from `5.6.7.8:9876` to `192.168.0.2:5678`, can Triber take advantage of it (i.e. be more available to other peers behind NAT without such a port forward) with enough config edits?

**Reply by qstokkink** [reply]
Whether all possible configurable settings of Tribler _should_ be accessible through the GUI is something to be discussed by @Tribler/dev, I can't offer an answer for this.

Regarding the external address: yes, there are network setups where port forwarding and interface specification helps you be "more available". However, unless you are not finding peers at all, I would not bother doing this because (a) it is annoying to manage forwarded ports, (b) in most cases this is unnecessary, and (c) this makes your computer a little bit easier to attack.

**Reply by kozlovsky** [reply]
Usually, [most users do not change default settings](https://service-design.co/95-of-the-people-stick-to-the-default-option-9e16316a64e1), and some options may look esoteric and obscure. Some options are added for experimental goals and are not important for the regular user. If we introduce the strict policy that the GUI settings window should display all options, the result may be more confusing to a regular user. Also, it would be harder to add experimental options.

Instead, we can have a documentation page dedicated to config settings that can be helpful for advanced users.

It may be helpful to have each option in `TriblerConfig` annotated with a `description` argument. Then, we can find a way to gather these `description` arguments and generate a documentation page from them.
